File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/ElectionQuestion.java
package com.github.dedis.popstellar.model.network.method.message.data.election;

import androidx.annotation.NonNull;

import com.github.dedis.popstellar.model.Immutable;
import com.github.dedis.popstellar.model.objects.Election;
import com.google.gson.annotations.SerializedName;

import java.util.*;

@Immutable
public class ElectionQuestion {

  private final String id;
  private final String question;

  @SerializedName(value = "voting_method")
  private final String votingMethod;

  @SerializedName(value = "ballot_options")
  private final List<String> ballotOptions;

  @SerializedName(value = "write_in")
  private final boolean writeIn;

  /** Constructor for a data Question, for the election setup */
  public ElectionQuestion(String electionId, Question question) {
    this.question = question.title;
    this.ballotOptions = Collections.unmodifiableList(question.ballotOptions);
    this.writeIn = question.writeIn;
    this.votingMethod = question.votingMethod;

    this.id = Election.generateElectionQuestionId(electionId, this.question);
  }

  public String getId() {
    return id;
  }

  public String getQuestion() {
    return question;
  }

  public boolean getWriteIn() {
    return writeIn;
  }

  public List<String> getBallotOptions() {
    return ballotOptions;
  }

  public String getVotingMethod() {
    return votingMethod;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ElectionQuestion that = (ElectionQuestion) o;
    return java.util.Objects.equals(getQuestion(), that.getQuestion())
        && java.util.Objects.equals(getId(), that.getId());
  }

  @Override
  public int hashCode() {
    return java.util.Objects.hash(
        getId(), getVotingMethod(), getWriteIn(), getBallotOptions(), getQuestion());
  }

  @NonNull
  @Override
  public String toString() {
    return "ElectionQuestion{"
        + "id='"
        + id
        + '\''
        + ", question='"
        + question
        + '\''
        + ", votingMethod='"
        + votingMethod
        + '\''
        + ", ballotOptions="
        + Arrays.toString(ballotOptions.toArray())
        + ", writeIn="
        + writeIn
        + '}';
  }

  /**
   * This data class holds the information of an ElectionQuestion except its id.
   *
   * <p>This is used to pack the question data when the election id is not available yet
   */
  @Immutable
  public static class Question {

    private final String title;
    private final String votingMethod;
    private final List<String> ballotOptions;
    private final boolean writeIn;

    public Question(
        String title, String votingMethod, List<String> ballotOptions, boolean writeIn) {
      this.title = title;
      this.votingMethod = votingMethod;
      this.ballotOptions = ballotOptions;
      this.writeIn = writeIn;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      Question that = (Question) o;
      return writeIn == that.writeIn
          && java.util.Objects.equals(title, that.title)
          && java.util.Objects.equals(votingMethod, that.votingMethod)
          && java.util.Objects.equals(ballotOptions, that.ballotOptions);
    }

    @Override
    public int hashCode() {
      return java.util.Objects.hash(title, votingMethod, ballotOptions, writeIn);
    }

    @NonNull
    @Override
    public String toString() {
      return "Question{"
          + "question='"
          + title
          + '\''
          + ", votingMethod='"
          + votingMethod
          + '\''
          + ", ballotOptions="
          + ballotOptions
          + ", writeIn="
          + writeIn
          + '}';
    }
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/ElectionQuestion.kt
package com.github.dedis.popstellar.model.network.method.message.data.election

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.model.objects.Election
import com.github.dedis.popstellar.utility.MessageValidator.verify
import com.google.gson.annotations.SerializedName
import java.util.Collections
import java.util.Objects

@Immutable
/** Constructor for a data Question, for the election setup */
class ElectionQuestion(electionId: String, question: Question) {
  val question: String
  val id: String

  @SerializedName(value = "voting_method") val votingMethod: String

  @SerializedName(value = "ballot_options") val ballotOptions: List<String>

  @SerializedName(value = "write_in") val writeIn: Boolean

  init {
    verify().isNotEmptyBase64(electionId, "election ID").validQuestions(listOf(question))

    this.question = question.title
    this.id = Election.generateElectionQuestionId(electionId, question.title)
    this.votingMethod = question.votingMethod
    this.ballotOptions = Collections.unmodifiableList(question.ballotOptions)
    this.writeIn = question.writeIn
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as ElectionQuestion
    return question == that.question && id == that.id
  }

  override fun hashCode(): Int {
    return Objects.hash(id, votingMethod, writeIn, ballotOptions, question)
  }

  override fun toString(): String {
    return "ElectionQuestion{id='$id', question='$question', votingMethod='$votingMethod', ballotOptions=${
      ballotOptions.toTypedArray().contentToString()
    }, writeIn=$writeIn}"
  }

  /**
   * This data class holds the information of an ElectionQuestion except its id.
   *
   * This is used to pack the question data when the election id is not available yet
   */
  @Immutable
  class Question(
    title: String,
    votingMethod: String,
    ballotOptions: List<String>,
    writeIn: Boolean,
  ) {
    val title: String
    val votingMethod: String
    val ballotOptions: List<String>
    val writeIn: Boolean

    init {
      verify().validQuestion(title, votingMethod, ballotOptions)

      this.title = title
      this.votingMethod = votingMethod
      this.ballotOptions = Collections.unmodifiableList(ballotOptions)
      this.writeIn = writeIn
    }

    override fun equals(other: Any?): Boolean {
      if (this === other) {
        return true
      }
      if (other == null || javaClass != other.javaClass) {
        return false
      }
      val that = other as Question
      return writeIn == that.writeIn &&
        title == that.title &&
        votingMethod == that.votingMethod &&
        ballotOptions == that.ballotOptions
    }

    override fun hashCode(): Int {
      return Objects.hash(title, votingMethod, ballotOptions, writeIn)
    }

    override fun toString(): String {
      return "Question{question='$title', votingMethod='$votingMethod', ballotOptions=$ballotOptions, writeIn=$writeIn}"
    }
  }
}
